homeassistant/components/upcloud/__init__.py

```diff
@@ -52,65 +64,200 @@
             {
                 vol.Required(CONF_USERNAME): cv.string,
                 vol.Required(CONF_PASSWORD): cv.string,
-                vol.Optional(CONF_SCAN_INTERVAL, default=SCAN_INTERVAL): cv.time_period,
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
             }
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass, config):
-    """Set up the UpCloud component."""
-    conf = config[DOMAIN]
-    username = conf.get(CONF_USERNAME)
-    password = conf.get(CONF_PASSWORD)
-    scan_interval = conf.get(CONF_SCAN_INTERVAL)
+class UpCloudDataUpdateCoordinator(
+    DataUpdateCoordinator[Dict[str, upcloud_api.Server]]
+):
+    """UpCloud data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistantType,
+        *,
+        cloud_manager: upcloud_api.CloudManager,
+        update_interval: timedelta,
+        username: str,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=f"{username}@UpCloud", update_interval=update_interval
+        )
+        self.cloud_manager = cloud_manager
+        self.unsub_handlers: List[CALLBACK_TYPE] = []
+
+    async def async_update_config(self, config_entry: ConfigEntry) -> None:
+        """Handle config update."""
+        self.update_interval = timedelta(
+            seconds=config_entry.options[CONF_SCAN_INTERVAL]
+        )
+
+    async def _async_update_data(self) -> Dict[str, upcloud_api.Server]:
+        return {
+            x.uuid: x
+            for x in await self.hass.async_add_executor_job(
```

Comment:
You're doing an `await` inside a for-loop. Why not do it in parallel using `asyncio.gather`?

Reply:
Rather than awaiting several calls in a loop, we're looping over the return value (list) of just one awaited method here.